docassemble/ALToolbox/data/static/phone-number-validation.js

```diff
@@ -1,26 +1,50 @@
+  *
+  * See phone_number_validation_demo.yml in this library for how to use this
+  *    feature of ALToolbox.
+  *
+  * We have discussed AsYouType input formatter, but haven't approached it yet.
+  *    Before discussing it, first see https://github.com/jackocnr/intl-tel-input/issues/346
+  *    where int-tel-input decided to get rid of their own version of auto
+  *    formatting user input and explained why. It's unclear if it's even possible.
+  *    One library to look into is https://nosir.github.io/cleave.js/. It's old,
+  *    but we might be able to use it as a model. We would want to test it
+  *    extensively with all the pitfalls that int-tel-input brings up and the
+  *    issues in the cleave repository itself.
+  *
+  * More Resources
+  *    1. What we're using: https://github.com/jackocnr/intl-tel-input
+  *    1. It uses https://www.npmjs.com/package/google-libphonenumber
+  *    1. https://github.com/google/libphonenumber/blob/master/FALSEHOODS.md
+  *    1. https://github.com/google/libphonenumber/blob/master/FAQ.md
+  */
   
-  // Set 'us' as the default if country_code has no value.
-  if (!country_code){
-     country_code = 'us'; 
-  }
-  // Loop thru all the phone input fields on the current screen
-  let phoneNodes = document.querySelectorAll( '.al-intl-phone' );  
-  for ( let node of phoneNodes ) {
-    let telObj = window.intlTelInput( node, {
-      initialCountry: country_code,        
-      utilsScript: "https://cdnjs.cloudflare.com/ajax/libs/intl-tel-input/17.0.3/js/utils.min.js"
-    });        
+  // Loop through all the .al_intl_phone input fields on the current screen
+  let phoneNodes = document.querySelectorAll( '.al_intl_phone' );  // Class given by the CustomDataType
+  for ( var node of phoneNodes ) {
+    var telObj = window.intlTelInput( node, {
+      // The default country without any input into the plugin is 'us'
+      // Once the user puts in the phone number of another country, though,
+      // it will remember that country
+      initialCountry: 'us',
+      utilsScript: "https://cdnjs.cloudflare.com/ajax/libs/intl-tel-input/17.0.15/js/utils.min.js"
+    });
     
-    // Save input as full number (eg +17024181234) to the backend        
-    $(node).val(telObj.getNumber());	
-
-    // Save country input text form (as opposed to digit) separately to the hidden field
-    // This value is "undefined" for US numbers, but our default value takes care of that.
-    $(country).val(telObj.getSelectedCountryData().iso2);      
+    // If the user already entered a number and comes back to the page,
+    // docassemble puts the previously saved value into the input field.
+    // Unfortunately, docassemble reformats the number and intl-tel-input
+    // thinks the number is invalid. It's complicated to describe, but you
+    // can see it if you go forward and back three times on our demo screens.
+    $(node).val(telObj.getNumber());
```

Comment:
Now I remember what happened regarding the hidden field: I originally used a new datatype: "country" and set a default value "us" on the screen, then in JS grabbed it and passed it to telObj. 

Later I felt since the field would be exposed to the developers, it's better to simplified the field setup and move some of the stuff to JS, and that's how the field became undefined. Thank you for catching it and discovering the fact that the hidden field is not needed anyway! 

Since in my branch I used the old version plug-in and it worked, I believe the key diff between your original old version and the new version is to save the input to the backend, thus I am adding comments before that line of code to provide some context.

Reply:
It doesn't actually save anything to the backend, but I can try to add a comment. I think this might be closer to what's going on:

> When the page first loads, gets the value from the input field, transforms it, puts the transformed value back into the input field, and guesses the country.

When a user hits 'back', docassemble automatically puts their previous input into the field, so intl-tel-input uses that. Docassemble is the only thing that saves anything to the backend. That happens when the user hits 'submit' to continue to the next page. This code doesn't do anything to the value when the user hits 'submit'.

Reply:
That text would probably be a good description of what the whole file is doing.

Reply:
Does that explanation make sense, @purplesky2016 ?

Reply:
Looks like we have different views. That's okay, it's just comments, you can do what you deem proper.